palette/src/xyz.rs

```diff
@@ -340,6 +340,38 @@ where
     }
 }
 
+impl<T> FromColorUnclamped<Okhsv<T>> for Xyz<D65, T>
+where
+    Okhsv<T>: IntoColorUnclamped<Oklab<T>>,
+    Self: FromColorUnclamped<Oklab<T>>,
+{
+    fn from_color_unclamped(color: Okhsv<T>) -> Self {
+        let oklab: Oklab<T> = color.into_color_unclamped();
+        Self::from_color_unclamped(oklab)
+    }
+}
+
+impl<T> FromColorUnclamped<Okhsl<T>> for Xyz<D65, T>
+where
+    Okhsl<T>: IntoColorUnclamped<Oklab<T>>,
+    Self: FromColorUnclamped<Oklab<T>>,
+{
+    fn from_color_unclamped(color: Okhsl<T>) -> Self {
+        let oklab: Oklab<T> = color.into_color_unclamped();
+        Self::from_color_unclamped(oklab)
+    }
+}
+
+impl<T> FromColorUnclamped<Okhwb<T>> for Xyz<D65, T>
+where
+    Okhwb<T>: IntoColorUnclamped<Okhsv<T>>,
+    Self: FromColorUnclamped<Okhsv<T>>,
+{
+    fn from_color_unclamped(color: Okhwb<T>) -> Self {
+        let okhsv: Okhsv<T> = color.into_color_unclamped();
+        Self::from_color_unclamped(okhsv)
+    }
+}
```

Comment:
These should not be necessary. This is precisely the type of code the macro should generate. Did you run into any issues with them? There may very well be a hole in its logic.

Reply:
I didn't run into issues. I just added stuff without any understanding until the compiler was satisfied. Then I stopped. Still without understanding :(
I will remove them.

Reply:
Let me know if the compiler complains without them (and without mentioning them in the exclusion list above), and we can figure it out. :slightly_smiling_face:

Reply:
I've removed them locally, and the compiler is fine.

Reply:
Nice! Then all should be fine if it's still possible to convert to `Xyz`. It's a bit tricky that it doesn't complain until you try to use it somewhere. There are a bunch of tests in `convert.rs` that checks this but it looks like the Ok* colors are not there yet.

Reply:
I  shouldn't have counted our chickens before they hatched. The compiler is fine, but when running tests, 3 doc tests fail. So I will leave them in.

Reply:
Can you paste the compiler output here, including hints and notes? I would like to see what it complains about.

Reply:
error[E0308]: mismatched types
   --> src\cast.rs:39:42
    |
6   | let color = cast::from_array::<Srgb<u8>>([23u8, 198]); // Too few components.
    |             ---------------------------- ^^^^^^^^^^^ expected an array with a fixed size of 3 elements, found one with 2 elements
    |             |
    |             arguments to this function are incorrect
    |
note: function defined here
   --> C:\Users\Navid\Softwareentwicklung\rust_projects\palette\palette\src\cast\array.rs:200:8
    |
200 | pub fn from_array<T>(array: T::Array) -> T
    |        ^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
warning: derive helper attribute is used before it is introduced
  --> src\convert.rs:220:3
   |
11 | #[palette(
   |   ^^^^^^^
...
15 | #[derive(PartialEq, Debug, FromColorUnclamped, WithAlpha)]
   |                            ------------------ the attribute is introduced here
   |
   = note: `#[warn(legacy_derive_helpers)]` on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #79202 <https://github.com/rust-lang/rust/issues/79202>

error[E0277]: the trait bound `Xyz: FromColorUnclamped<Okhwb>` is not satisfied
  --> src\convert.rs:224:28
   |
15 | #[derive(PartialEq, Debug, FromColorUnclamped, WithAlpha)]
   |                            ^^^^^^^^^^^^^^^^^^ the trait `FromColorUnclamped<Okhwb>` is not implemented for `Xyz`
   |
   = help: the following other types implement trait `FromColorUnclamped<T>`:
             <Xyz as FromColorUnclamped<CssRgb>>
             <Xyz<D65, T> as FromColorUnclamped<Oklab<T>>>
             <Xyz<D65, T> as FromColorUnclamped<Oklch<T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Alpha<_C, _A>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsl<_S, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsluv<Wp, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsv<_S, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hwb<_S, T>>>
           and 8 others
   = note: required because of the requirements on the impl of `FromColorUnclamped<Okhwb>` for `Rgb`
   = help: see issue #48214
   = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable
   = note: this error originates in the derive macro `FromColorUnclamped` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `Xyz: FromColorUnclamped<Okhsl>` is not satisfied
  --> src\convert.rs:224:28
   |
15 | #[derive(PartialEq, Debug, FromColorUnclamped, WithAlpha)]
   |                            ^^^^^^^^^^^^^^^^^^ the trait `FromColorUnclamped<Okhsl>` is not implemented for `Xyz`
   |
   = help: the following other types implement trait `FromColorUnclamped<T>`:
             <Xyz as FromColorUnclamped<CssRgb>>
             <Xyz<D65, T> as FromColorUnclamped<Oklab<T>>>
             <Xyz<D65, T> as FromColorUnclamped<Oklch<T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Alpha<_C, _A>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsl<_S, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsluv<Wp, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsv<_S, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hwb<_S, T>>>
           and 8 others
   = note: required because of the requirements on the impl of `FromColorUnclamped<Okhsl>` for `Rgb`
   = help: see issue #48214
   = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable
   = note: this error originates in the derive macro `FromColorUnclamped` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `Xyz: FromColorUnclamped<Okhsv>` is not satisfied
  --> src\convert.rs:224:28
   |
15 | #[derive(PartialEq, Debug, FromColorUnclamped, WithAlpha)]
   |                            ^^^^^^^^^^^^^^^^^^ the trait `FromColorUnclamped<Okhsv>` is not implemented for `Xyz`
   |
   = help: the following other types implement trait `FromColorUnclamped<T>`:
             <Xyz as FromColorUnclamped<CssRgb>>
             <Xyz<D65, T> as FromColorUnclamped<Oklab<T>>>
             <Xyz<D65, T> as FromColorUnclamped<Oklch<T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Alpha<_C, _A>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsl<_S, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsluv<Wp, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsv<_S, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hwb<_S, T>>>
           and 8 others
   = note: required because of the requirements on the impl of `FromColorUnclamped<Okhsv>` for `Rgb`
   = help: see issue #48214
   = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable
   = note: this error originates in the derive macro `FromColorUnclamped` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 3 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0277`.
Couldn't compile the test.

error[E0277]: the trait bound `Xyz: FromColorUnclamped<Okhwb>` is not satisfied
 --> src\convert.rs:91:28
  |
7 | #[derive(PartialEq, Debug, FromColorUnclamped)]
  |                            ^^^^^^^^^^^^^^^^^^ the trait `FromColorUnclamped<Okhwb>` is not implemented for `Xyz`
  |
  = help: the following other types implement trait `FromColorUnclamped<T>`:
            <Xyz as FromColorUnclamped<Xyz100>>
            <Xyz<D65, T> as FromColorUnclamped<Oklab<T>>>
            <Xyz<D65, T> as FromColorUnclamped<Oklch<T>>>
            <Xyz<Wp, T> as FromColorUnclamped<Alpha<_C, _A>>>
            <Xyz<Wp, T> as FromColorUnclamped<Hsl<_S, T>>>
            <Xyz<Wp, T> as FromColorUnclamped<Hsluv<Wp, T>>>
            <Xyz<Wp, T> as FromColorUnclamped<Hsv<_S, T>>>
            <Xyz<Wp, T> as FromColorUnclamped<Hwb<_S, T>>>
          and 8 others
  = help: see issue #48214
  = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable
  = note: this error originates in the derive macro `FromColorUnclamped` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `Xyz: FromColorUnclamped<Okhsl>` is not satisfied
 --> src\convert.rs:91:28
  |
7 | #[derive(PartialEq, Debug, FromColorUnclamped)]
  |                            ^^^^^^^^^^^^^^^^^^ the trait `FromColorUnclamped<Okhsl>` is not implemented for `Xyz`
  |
  = help: the following other types implement trait `FromColorUnclamped<T>`:
            <Xyz as FromColorUnclamped<Xyz100>>
            <Xyz<D65, T> as FromColorUnclamped<Oklab<T>>>
            <Xyz<D65, T> as FromColorUnclamped<Oklch<T>>>
            <Xyz<Wp, T> as FromColorUnclamped<Alpha<_C, _A>>>
            <Xyz<Wp, T> as FromColorUnclamped<Hsl<_S, T>>>
            <Xyz<Wp, T> as FromColorUnclamped<Hsluv<Wp, T>>>
            <Xyz<Wp, T> as FromColorUnclamped<Hsv<_S, T>>>
            <Xyz<Wp, T> as FromColorUnclamped<Hwb<_S, T>>>
          and 8 others
  = help: see issue #48214
  = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable
  = note: this error originates in the derive macro `FromColorUnclamped` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `Xyz: FromColorUnclamped<Okhsv>` is not satisfied
 --> src\convert.rs:91:28
  |
7 | #[derive(PartialEq, Debug, FromColorUnclamped)]
  |                            ^^^^^^^^^^^^^^^^^^ the trait `FromColorUnclamped<Okhsv>` is not implemented for `Xyz`
  |
  = help: the following other types implement trait `FromColorUnclamped<T>`:
            <Xyz as FromColorUnclamped<Xyz100>>
            <Xyz<D65, T> as FromColorUnclamped<Oklab<T>>>
            <Xyz<D65, T> as FromColorUnclamped<Oklch<T>>>
            <Xyz<Wp, T> as FromColorUnclamped<Alpha<_C, _A>>>
            <Xyz<Wp, T> as FromColorUnclamped<Hsl<_S, T>>>
            <Xyz<Wp, T> as FromColorUnclamped<Hsluv<Wp, T>>>
            <Xyz<Wp, T> as FromColorUnclamped<Hsv<_S, T>>>
            <Xyz<Wp, T> as FromColorUnclamped<Hwb<_S, T>>>
          and 8 others
  = help: see issue #48214
  = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable
  = note: this error originates in the derive macro `FromColorUnclamped` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0277`.
Couldn't compile the test.

error[E0369]: cannot add `Rgb<Srgb, {float}>` to `Rgb<Srgb, {float}>`
 --> src\lib.rs:26:37
  |
9 | let whatever_it_becomes = orangeish + blueish; // Does not compile
  |                           --------- ^ ------- Rgb<Srgb, {float}>
  |                           |
  |                           Rgb<Srgb, {float}>

error: aborting due to previous error

For more information about this error, try `rustc --explain E0369`.
error[E0277]: the trait bound `Xyz: FromColorUnclamped<Okhwb>` is not satisfied
  --> src\lib.rs:1772:10
   |
12 | #[derive(FromColorUnclamped, WithAlpha)]
   |          ^^^^^^^^^^^^^^^^^^ the trait `FromColorUnclamped<Okhwb>` is not implemented for `Xyz`
   |
   = help: the following other types implement trait `FromColorUnclamped<T>`:
             <Xyz as FromColorUnclamped<Color>>
             <Xyz<D65, T> as FromColorUnclamped<Oklab<T>>>
             <Xyz<D65, T> as FromColorUnclamped<Oklch<T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Alpha<_C, _A>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsl<_S, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsluv<Wp, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsv<_S, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hwb<_S, T>>>
           and 8 others
   = note: required because of the requirements on the impl of `FromColorUnclamped<Okhwb>` for `Rgb`
   = help: see issue #48214
   = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable
   = note: this error originates in the derive macro `FromColorUnclamped` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `Xyz: FromColorUnclamped<Okhsl>` is not satisfied
  --> src\lib.rs:1772:10
   |
12 | #[derive(FromColorUnclamped, WithAlpha)]
   |          ^^^^^^^^^^^^^^^^^^ the trait `FromColorUnclamped<Okhsl>` is not implemented for `Xyz`
   |
   = help: the following other types implement trait `FromColorUnclamped<T>`:
             <Xyz as FromColorUnclamped<Color>>
             <Xyz<D65, T> as FromColorUnclamped<Oklab<T>>>
             <Xyz<D65, T> as FromColorUnclamped<Oklch<T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Alpha<_C, _A>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsl<_S, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsluv<Wp, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsv<_S, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hwb<_S, T>>>
           and 8 others
   = note: required because of the requirements on the impl of `FromColorUnclamped<Okhsl>` for `Rgb`
   = help: see issue #48214
   = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable
   = note: this error originates in the derive macro `FromColorUnclamped` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `Xyz: FromColorUnclamped<Okhsv>` is not satisfied
  --> src\lib.rs:1772:10
   |
12 | #[derive(FromColorUnclamped, WithAlpha)]
   |          ^^^^^^^^^^^^^^^^^^ the trait `FromColorUnclamped<Okhsv>` is not implemented for `Xyz`
   |
   = help: the following other types implement trait `FromColorUnclamped<T>`:
             <Xyz as FromColorUnclamped<Color>>
             <Xyz<D65, T> as FromColorUnclamped<Oklab<T>>>
             <Xyz<D65, T> as FromColorUnclamped<Oklch<T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Alpha<_C, _A>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsl<_S, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsluv<Wp, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hsv<_S, T>>>
             <Xyz<Wp, T> as FromColorUnclamped<Hwb<_S, T>>>
           and 8 others
   = note: required because of the requirements on the impl of `FromColorUnclamped<Okhsv>` for `Rgb`
   = help: see issue #48214
   = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable
   = note: this error originates in the derive macro `FromColorUnclamped` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0277`.
Couldn't compile the test.

error: cannot find macro `lazy_select` in this scope
 --> src\macros\lazy_select.rs:4:14
  |
3 | let result = lazy_select! {
  |              ^^^^^^^^^^^

error: aborting due to previous error

Couldn't compile the test.

Reply:
Thanks! Too bad it doesn't say a lot, but I hoped it would give a hint. Anyway, I can see if I can fix it later. You gave me access after all :+1:

Reply:
Damn, I forgot to investigate why these are needed... 😬